File: advent_of_code/2023/13/day13.hs
import Data.Char (isSpace)
import Data.List (isPrefixOf, singleton, intercalate)
import Data.Functor ((<&>))
import Data.Maybe (fromMaybe)

type Grid = [[String]]

indexOf :: Eq a => [a] -> [a] -> Maybe Int
indexOf [] _ = Nothing
indexOf a b = if b `isPrefixOf` a then Just 0 else indexOf (tail a) b <&> (+1)

split :: String -> String -> [String]
split [] _ = []
split str sep = case str `indexOf` sep of
                  Just i -> take i str : split (drop (i + length sep) str) sep
                  Nothing -> [str]

trim :: String -> String
trim = reverse . dropWhile isSpace . reverse . dropWhile isSpace

readInput :: String -> [Grid]
readInput s = split s "\n\n" <&> \g -> filter (not . null) (split g "\n") <&> map singleton

dumpGrids :: [Grid] -> String
dumpGrids = concatMap dumpGrid
  where
    dumpGrid :: Grid -> String
    dumpGrid grid = (intercalate "\n" . fmap concat $ grid) ++ "\n\n"

range :: Int -> Int -> [Int]
range i n
  | i >= n = []
  | otherwise = i : range (i + 1) n

count :: (a -> Bool) -> [a] -> Int
count _ [] = 0
count f (a:as) = if f a then 1 + count f as else count f as

get :: [a] -> Maybe a
get [] = Nothing
get (a:_) = Just a

mirrorRow :: Grid -> Maybe Int
mirrorRow grid = get $ filter numDifferences (range 1 (length grid))
  where
    numDifferences :: Int -> Bool
    numDifferences i = all (and . uncurry (zipWith (==))) (zip (reverse . take i $ grid) (drop i grid))

getOrElse :: a -> Maybe a -> a
getOrElse _ (Just a) = a
getOrElse a Nothing = a

columns :: Grid -> Grid
columns grid = range 0 (length . head $ grid) <&> \i -> fmap (!! i) grid


part1 :: [Grid] -> Int
part1 = sum . fmap getMirrorRows
  where
    getMirrorRows grid = case mirrorRow grid of
      Just r -> 100 * r
      Nothing -> fromMaybe 0 (mirrorRow . columns $ grid)


mirrorRowWithSmudges :: Grid -> Maybe Int
mirrorRowWithSmudges grid = get $ filter numDifferences (range 1 (length grid))
  where
    numDifferences :: Int -> Bool
    numDifferences i = 1 == sum (zip (reverse . take i $ grid) (drop i grid) <&> count id . uncurry (zipWith (/=)))

part2 :: [Grid] -> Int
part2 = sum . fmap getMirrorRows
  where
    getMirrorRows grid = case mirrorRowWithSmudges grid of
      Just r -> 100 * r
      Nothing -> fromMaybe 0 (mirrorRowWithSmudges . columns $ grid)


test_input = "\n\
\#.##..##.\n\
\..#.##.#.\n\
\##......#\n\
\##......#\n\
\..#.##.#.\n\
\..##..##.\n\
\#.#.##.#.\n\
\\n\
\#...##..#\n\
\#....#..#\n\
\..##..###\n\
\#####.##.\n\
\#####.##.\n\
\..##..###\n\
\#....#..#\n\
\\n"

main :: IO ()
main = do
  let filename = "input.txt"
  let debug = False
  input <-
    if debug then return test_input
             else readFile filename
  let grids = readInput input
  -- putStrLn $ "Grids: \n" ++ dumpGrids grids
  putStrLn $ "2023 Day 13, Part 1: " ++ show (part1 grids)
  putStrLn $ "2023 Day 13, Part 2: " ++ show (part2 grids)

File: scala/aoc/src/main/scala/aoc2023/day13.scala
package aoc2023

object Day13 {
  type Grid = List[List[String]]
  def readInput(input: String): List[Grid] = {
    input.split("\n\n").map {
      grid =>
        grid.trim.split("\n").map(_.trim).filter(_.nonEmpty)
          .map(line => line.map(_.toString).toList).toList
    }.toList
  }

  def mirrorRow(grid: Grid): Int = {
    (1 until grid.length).filter { i =>
      grid.take(i).reverse.zip(grid.drop(i)).forall {
        (row1, row2) =>
          row1.zip(row2).forall { (c1, c2) => c1 == c2 }
      }
    }.headOption.getOrElse(0)
  }

  def mirrorRowWithSmudges(grid: Grid): Option[Int] = {
    (1 until grid.length).filter { i =>
      val numDifferences = grid.take(i).reverse.zip(grid.drop(i)).map { 
        (row1, row2) =>
          row1.zip(row2).count { (c1, c2) => c1 != c2 }
      }.sum
      numDifferences == 1
    }.headOption
  }

  def columns(grid: Grid): Grid = {
    grid(0).indices.map { i =>
      grid.map(line => line(i))
    }.toList
  }

  def part1(input: String): Int = {
    val grids = readInput(input)
    grids.map {
      grid =>
        100 * mirrorRow(grid) + mirrorRow(columns(grid))
    }.sum
  }

  def part2(input: String): Int = {
    val grids = readInput(input)
    grids.map {
      grid =>
        mirrorRowWithSmudges(grid).map(_ * 100)
          .orElse(mirrorRowWithSmudges(columns(grid)))
          .getOrElse(0)
    }.sum
  }

  def main(args: Array[String]): Unit = {
    val filename = "input.txt"
    val test_input = """
      #.##..##.
      ..#.##.#.
      ##......#
      ##......#
      ..#.##.#.
      ..##..##.
      #.#.##.#.

      #...##..#
      #....#..#
      ..##..###
      #####.##.
      #####.##.
      ..##..###
      #....#..#
    """
    val debug = false
    val input = 
      if (debug) test_input 
      else io.Source.fromFile(filename).getLines().mkString("\n")
    
    println(s"2023 Day 13, Part 1: ${part1(input)}")
    println(s"2023 Day 13, Part 2: ${part2(input)}")
  }
}
